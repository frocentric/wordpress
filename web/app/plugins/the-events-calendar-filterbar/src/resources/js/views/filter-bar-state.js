/* global tribe, tribe_events_filter_bar_js_config */
/* eslint-disable no-var, strict */

/**
 * Makes sure we have all the required levels on the Tribe Object.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar = tribe.filterBar || {};

/**
 * Configures Filter Bar State Object in the Global Tribe variable.
 *
 * @since 5.0.0
 *
 * @type   {PlainObject}
 */
tribe.filterBar.filterBarState = {};

/**
 * Initializes in a Strict env the code that manages the filter bar state.
 *
 * @since 5.0.0
 *
 * @param  {PlainObject} $   jQuery
 * @param  {PlainObject} obj tribe.filterBar.filterBarState
 *
 * @return {void}
 */
( function( $, obj ) {
	'use strict';

	/**
	 * Selectors used for configuration and setup.
	 *
	 * @since 5.0.0
	 *
	 * @type {PlainObject}
	 */
	obj.selectors = {
		actionDone: '[data-js="tribe-filter-bar__action-done"]',
		container: '[data-js="tribe-events-view"]', // @todo [BTRIA-621]: @paulmskim remove when proper hooks are in place.
		dataScript: '[data-js="tribe-events-view-data"]', // @todo [BTRIA-621]: @paulmskim remove when proper hooks are in place.
		filterBar: '[data-js~="tribe-filter-bar"]',
		filterBarOpen: '.tribe-filter-bar--open',
		filterButton: '[data-js~="tribe-events-filter-button"]',
		filterButtonActive: '.tribe-events-c-events-bar__filter-button--active',
		filterButtonText: '.tribe-events-c-events-bar__filter-button-text',
	};

	/**
	 * Open filter bar.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.openFilterBar = function( $container ) {
		var $filterButton = $container.find( obj.selectors.filterButton );
		var $filterButtonText = $filterButton.find( obj.selectors.filterButtonText );
		var $actionDone = $container.find( obj.selectors.actionDone );
		var $filterBar = $container.find( obj.selectors.filterBar );

		$filterButton.addClass( obj.selectors.filterButtonActive.className() );
		$filterButtonText.text( tribe_events_filter_bar_js_config.l10n.hide_filters );
		$filterBar.addClass( obj.selectors.filterBarOpen.className() );

		tribe.events.views.accordion.setOpenAccordionA11yAttrs( $filterButton, $filterBar );
		tribe.events.views.accordion.setOpenAccordionA11yAttrs( $actionDone, $filterBar );
	};

	/**
	 * Close filter bar.
	 *
	 * @since  5.0.0
	 *
	 * @param  {jQuery} $container jQuery object of view container.
	 *
	 * @return {void}
	 */
	obj.closeFilterBar = function( $container ) {
		var $filterButton = $container.find( obj.selectors.filterButton );
		var $filterButtonText = $filterButton.find( obj.selectors.filterButtonText );
		var $actionDone = $container.find( obj.selectors.actionDone );
		var $filterBar = $container.find( obj.selectors.filterBar );

		$filterButton.removeClass( obj.selectors.filterButtonActive.className() );
		$filterButtonText.text( tribe_events_filter_bar_js_config.l10n.show_filters );
		$filterBar.removeClass( obj.selectors.filterBarOpen.className() );

		tribe.events.views.accordion.setCloseAccordionA11yAttrs( $filterButton, $filterBar );
		tribe.events.views.accordion.setCloseAccordionA11yAttrs( $actionDone, $filterBar );
	};

	/**
	 * Setup filter bar.
	 *
	 * @since  5.0.2
	 *
	 * @param  {HTMLElement} container HTML element of the container containing the script tag calling setup
	 *
	 * @return {void}
	 */
	obj.setup = function( container ) {
		// @todo [BTRIA-621]: @paulmskim all of this is temporary, this will hook into the breakpoints.js in TEC when released with TEC containing the fixes.
		var $container = $( container );

		if ( ! $container.is( obj.selectors.container ) ) {
			return;
		}

		var $filterBar = $container.find( obj.selectors.filterBar );

		// Return early if mobile initial state control is false.
		if ( ! $filterBar.data( 'mobileInitialStateControl' ) ) {
			return;
		}

		var $data = $container.find( obj.selectors.dataScript );
		var data = {};

		// If we have data element set it up.
		if ( $data.length ) {
			data = JSON.parse( $.trim( $data.text() ) );
		}

		var breakpoints = Object.keys( data.breakpoints );

		breakpoints.forEach( function( breakpoint ) {
			if ( 'medium' === breakpoint && $container.outerWidth() < data.breakpoints[ breakpoint ] ) {
				obj.closeFilterBar( $container );
			}
		} );
	};
} )( jQuery, tribe.filterBar.filterBarState );
