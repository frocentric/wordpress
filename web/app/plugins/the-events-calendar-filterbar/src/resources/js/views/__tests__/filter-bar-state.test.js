describe( 'Filter Bar State', () => {
	beforeAll( () => {
		String.prototype.className = function() {
			if (
				(
					'string' !== typeof this &&
					! this instanceof String /* eslint-disable-line no-unsafe-negation */
				) ||
				'function' !== typeof this.replace
			) {
				return this;
			}

			return this.replace( '.', '' );
		};

		global.tribe = {};
		require( '../filter-bar-state' );
		tribe.events = {
			views: {
				accordion: {},
			},
		};
	} );

	afterAll( () => {
		delete String.prototype.className;
		delete global.tribe;
	} );

	beforeEach( () => {
		global.tribe_events_filter_bar_js_config = {
			l10n: {
				hide_filters: 'Hide filters',
			},
		};
	} );

	afterEach( () => {
		delete global.tribe_events_filter_bar_js_config;
	} );

	describe( 'Selectors', () => {
		test( 'Should match snapshot', () => {
			const selectors = JSON.stringify( tribe.filterBar.filterBarState.selectors );
			expect( selectors ).toMatchSnapshot();
		} );
	} );

	describe( 'Open filter bar', () => {
		test( 'Should open filter bar', () => {
			// Setup test.
			tribe.events.views.accordion.setOpenAccordionA11yAttrs = jest.fn();
			const $container = $();
			const $filterButton = $();
			const $filterButtonText = $();
			const $filterBar = $();
			$container.find = ( selector ) => {
				switch ( selector ) {
					case tribe.filterBar.filterBarState.selectors.filterButton:
						return $filterButton;
					case tribe.filterBar.filterBarState.selectors.filterBar:
						return $filterBar;
					default:
						return $();
				}
			};
			$filterButton.find = () => $filterButtonText;
			$filterButton.addClass = jest.fn();
			$filterButtonText.text = jest.fn();
			$filterBar.addClass = jest.fn();

			// Test.
			tribe.filterBar.filterBarState.openFilterBar( $container );

			// Confirm final states.
			expect( tribe.events.views.accordion.setOpenAccordionA11yAttrs.mock.calls.length ).toBe( 2 );
			expect( $filterButton.addClass.mock.calls.length ).toBe( 1 );
			expect( $filterButtonText.text.mock.calls.length ).toBe( 1 );
			expect( $filterBar.addClass.mock.calls.length ).toBe( 1 );
		} );
	} );

	describe( 'Close filter bar', () => {
		test( 'Should close filter bar', () => {
			tribe.events.views.accordion.setCloseAccordionA11yAttrs = jest.fn();
			const $container = $();
			const $filterButton = $();
			const $filterButtonText = $();
			const $filterBar = $();
			$container.find = ( selector ) => {
				switch ( selector ) {
					case tribe.filterBar.filterBarState.selectors.filterButton:
						return $filterButton;
					case tribe.filterBar.filterBarState.selectors.filterBar:
						return $filterBar;
					default:
						return $();
				}
			};
			$filterButton.find = () => $filterButtonText;
			$filterButton.removeClass = jest.fn();
			$filterButtonText.text = jest.fn();
			$filterBar.removeClass = jest.fn();

			// Test.
			tribe.filterBar.filterBarState.closeFilterBar( $container );

			// Confirm final states.
			expect( tribe.events.views.accordion.setCloseAccordionA11yAttrs.mock.calls.length ).toBe( 2 );
			expect( $filterButton.removeClass.mock.calls.length ).toBe( 1 );
			expect( $filterButtonText.text.mock.calls.length ).toBe( 1 );
			expect( $filterBar.removeClass.mock.calls.length ).toBe( 1 );
		} );
	} );
} );
