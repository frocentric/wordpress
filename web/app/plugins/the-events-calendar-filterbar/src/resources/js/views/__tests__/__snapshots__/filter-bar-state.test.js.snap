// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter Bar State Selectors Should match snapshot 1`] = `"{\\"actionDone\\":\\"[data-js=\\\\\\"tribe-filter-bar__action-done\\\\\\"]\\",\\"container\\":\\"[data-js=\\\\\\"tribe-events-view\\\\\\"]\\",\\"dataScript\\":\\"[data-js=\\\\\\"tribe-events-view-data\\\\\\"]\\",\\"filterBar\\":\\"[data-js~=\\\\\\"tribe-filter-bar\\\\\\"]\\",\\"filterBarOpen\\":\\".tribe-filter-bar--open\\",\\"filterButton\\":\\"[data-js~=\\\\\\"tribe-events-filter-button\\\\\\"]\\",\\"filterButtonActive\\":\\".tribe-events-c-events-bar__filter-button--active\\",\\"filterButtonText\\":\\".tribe-events-c-events-bar__filter-button-text\\"}"`;
